activerecord/lib/active_record/attribute_set.rb

```diff
@@ -64,9 +64,7 @@ def freeze
     end
 
     def deep_dup
-      dup.tap do |copy|
-        copy.instance_variable_set(:@attributes, attributes.deep_dup)
-      end
+      self.class.new(attributes.deep_dup)
```

Comment:
Yes, this change is to avoid the wasted shallow dup in `initialize_dup`. 

I think using [`Class#allocate`](https://ruby-doc.org/core-2.4.1/Class.html#method-i-allocate) may be the best approach:

```ruby
def deep_dup
  self.class.allocate.tap do |copy|
    copy.instance_variable_set(:@attributes, attributes.deep_dup)
  end
end
```

That seems to do exactly what we want: create a new instance of `AttributeSet` without calling `initialize`, so that we can initialize `@attributes` (and any future instance variables) ourselves within `deep_dup` and avoid wasted shallow duplication. (This is effectively what `Object#dup` does under the hood.) 

As you would expect, this variation provides the same speedup as the current optimization.